From ec4bffd288d0c9736a08f7e3f3b4f98dc43b3354 Mon Sep 17 00:00:00 2001
From: Andy Yan <andy.yan@rock-chips.com>
Date: Wed, 30 Dec 2020 10:58:24 +0800
Subject: [PATCH] drm/rockchip: vop2: Fix the max_input/output width of cluster
 sub window

According to TRM, the max_intput/output width of a cluster window
should reduce by half when work in two win mode.

Change-Id: I5cbe10605e8895eb307df9591b222c9293bee21a
Signed-off-by: Andy Yan <andy.yan@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_vop2.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
index 1db0fa0857f7..67dc6801f013 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
@@ -4345,12 +4345,16 @@ static int vop2_plane_init(struct vop2 *vop2, struct vop2_win *win, unsigned lon
 	vop2_plane_create_feature_property(vop2, win);
 	max_width = vop2->data->max_input.width;
 	max_height = vop2->data->max_input.height;
+	if (win->feature & WIN_FEATURE_CLUSTER_SUB)
+		max_width >>= 1;
 	win->input_width_prop = drm_property_create_range(vop2->drm_dev, DRM_MODE_PROP_IMMUTABLE,
 							  "INPUT_WIDTH", 0, max_width);
 	win->input_height_prop = drm_property_create_range(vop2->drm_dev, DRM_MODE_PROP_IMMUTABLE,
 							   "INPUT_HEIGHT", 0, max_height);
 	max_width = vop2->data->max_output.width;
 	max_height = vop2->data->max_output.height;
+	if (win->feature & WIN_FEATURE_CLUSTER_SUB)
+		max_width >>= 1;
 	win->output_width_prop = drm_property_create_range(vop2->drm_dev, DRM_MODE_PROP_IMMUTABLE,
 							   "OUTPUT_WIDTH", 0, max_width);
 	win->output_height_prop = drm_property_create_range(vop2->drm_dev, DRM_MODE_PROP_IMMUTABLE,
-- 
2.35.3

